### ISSUE TITLE
Tools such as ocamldoc, ocamllex, … are not compiled using ocamlopt.opt when available

### ISSUE THREAD

**Issue by kit-ty-kate** [ocaml/ocaml#12921]
Behaviour observed on the 4.14, 5.0, 5.1 (except ocamldoc was fine) and trunk branches as of today:
```
$ ./configure && make -j10
[...]
../../ocamlopt.opt -nostdlib -I ../../stdlib -I ../../otherlibs/unix -linkall -a -cclib -lthreadsnat -o threads.cmxa thread.cmx event.cmx
[...]
./boot/ocamlrun ./ocamlopt -nostdlib -I ./stdlib -I otherlibs/dynlink -g -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48 -warn-error +a -bin-annot -strict-formats -I ocamldoc -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I driver -I toplevel -I tools -I runtime -I otherlibs/dynlink -I otherlibs/str -I otherlibs/systhreads -I otherlibs/unix -I otherlibs/runtime_events  -c ocamldoc/odoc_config.ml
```

On slow computers (such as when using an emulator), the difference between using `./boot/ocamlrun ./ocamlopt` and `../../ocamlopt.opt` is significant enough to be noticable.

**Comment by gasche** [comment]
There may be a regression in this respect caused by the (overall positive) changes to avoid recursive Makefile invocations by moving all rules to the root Makefile. Before (say, in 5.0) `ocamldoc` had a Makefile that used BEST_OCAMLC and BEST_OCAMLOPT, computed at the time the `ocamldoc` make is invoked. Now that the makefiles are merged:

1. I cannot find a trace of use of BEST_OCAMLOPT in the root Makefile, so I don't think that they are ever used for ocamldoc, ocamllex etc. (It remains used in otherlibs/ which still has a separate makefile)
2. BEST_OCAMLC is used, but it may be computed at the time the root Makefile is invoked, which is too soon as ocamlc.opt has not been built yet at this point. (Or maybe this magically works because of obscure lazy evaluation rules of Makefile, I don't know!)

cc @dra27 and @shindere who know about this

**Comment by dra27** [comment]
I've put this in the 5.2 milestone, because there is a definite workaround to recover the slowness for ocamldoc. The larger "plan" for this is better executed after the "otherlibs" Makefiles are merged.

**Comment by shindere** [comment]
#13048 has just been opened to fix this.

The fix's purpose is to address the recent regressions: it makes sure
both ocamldoc.opt and ocamltest.opt are compiled with the best native
compiler available but it does not implement this for ocamllex.opt or
the native tools because it seems this was never done that way in the
past (at least, it was not done in 4.13 either). So the purpose of the
fix proposed for this issue is to bring the situation back to what it
used to be, in preparation for the upcoming beta version of OCaml5.2.
The situation can be improved later.

**Comment by shindere** [comment]
You were right, @gasche. Thanks.

**Comment by gasche** [comment]
This is now fixed in #13048 (and cherry-picked in the 5.2 dev branch). Thanks @kit-ty-kate for the report and @shindere for the fix.
